core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,378 @@
+                cancel();
+                return;
+            }
+
+            if (canceled) {
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                handleRequest(n);
+            } else {
+                executor.execute(() -> handleRequest(n));
+            }
+        }
+
+        private void handleRequest(long n) {
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            flush();
+
+            final long toRequest = maxConcurrency - childSubscribers.size();
+            if (toRequest > 0) {
+                upstream.request(toRequest);
+            }
+
+            requestAllAvailable();
+        }
+
+        @Override
+        public void cancel() {
+            upstream.cancel();
+            cancelChildSubscribers();
+        }
+
+        private void cancelChildSubscribers() {
+            childSubscribers.forEach(FlatMapSubscriber::cancel);
+        }
+
+        void childSubscribed(FlatMapSubscriber<T, U> child) {
+            pendingSubscriptions--;
+            childSubscribers.add(child);
+
+            requestAllAvailable();
+        }
+
+        private long getAvailableBufferSpace() {
+            if (requestedByDownstream == Long.MAX_VALUE) {
+                return Long.MAX_VALUE;
+            }
+
+            final Optional<Long> requested = childSubscribers.stream().map(FlatMapSubscriber::getRequested)
+                                                             .reduce(LongMath::saturatedAdd);
+
+            return maxConcurrency - requested.orElse(0L) - buffer.size();
+        }
+
+        private void requestAllAvailable() {
+            if (childSubscribers.isEmpty()) {
+                return;
+            }
+
+            final long available = getAvailableBufferSpace();
+
+            if (available == Long.MAX_VALUE) {
+                childSubscribers.forEach(sub -> sub.request(Long.MAX_VALUE));
+                return;
+            }
+
+            final List<FlatMapSubscriber<T, U>> toRequest = childSubscribers.stream()
+                                                                            .filter(sub -> sub.getRequested() ==
+                                                                                           0)
+                                                                            .limit(available)
+                                                                            .collect(toImmutableList());
+            toRequest.forEach(sub -> sub.request(1));
+        }
+
+        private void flush() {
+            while (requestedByDownstream > 0 && !buffer.isEmpty()) {
+                final U value = buffer.remove();
+
+                publishDownstream(value);
+            }
+        }
+
+        void completeChild(FlatMapSubscriber<T, U> child) {
+            childSubscribers.remove(child);
+
+            if (childSubscribers.isEmpty() && pendingSubscriptions == 0 && completing) {
```

Comment:
Is it possible that the queue is non-empty even though this condition is true?
It seems like we should also check if the queue is drained before checking this condition. (ditto for the other condition at `onComplete`)

I think we may also need to check this condition at `handleRequest` as well.

```suggestion
            if (childSubscribers.isEmpty() && pendingSubscriptions == 0 && completing && buffer.empty()) {
```